Hometask 2 3 4

anton-kuvaldin opened the pull request:
Был вынужден во многом обращаться за "вдохновением" к чужим кодам. Подумал, что лучше так и выполнить задание, чем не выполнить вообще. Если вас обеспокоит "похожесть" кода, смело вините только меня.
Большую часть этих изменений я закоммитил ещё неделю назад, но закоммитил в ветку hometask-1, поэтому вы не увидели этих изменений(можно в истории посмотреть). В этой ветке первым коммитом откатил изменения назад и следующим добавил изменения(домашки 2, 3 и 4)

anton-kuvaldin committed:
Откат изменений к первой версии(home-task-1)

Запутался и запушил домашки 2 и 3 в ветку с заданием 1. Откатываю изменения коммитом в новую ветку для заданий 2, 3 и 4

anton-kuvaldin committed:
Домашки 2, 3 и 4

Частично выполненые домашние задания 2, 3 и 4

anton-kuvaldin committed:
Merge branch 'hometask-2-3-4' of https://github.com/megahard25/smart-home-2020 into hometask-2-3-4

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -19,49 +27,7 @@ public static void main(String... args) throws IOException {
         SensorEvent event = getNextSensorEvent();
         while (event != null) {
```
Перенесите логику цикла обработки событий в отдельный класс https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%BF%D0%B5%D1%80%D0%B5%D0%BD%D0%B5%D1%81%D0%B8%D1%82%D0%B5-%D0%BB%D0%BE%D0%B3%D0%B8%D0%BA%D1%83-%D1%86%D0%B8%D0%BA%D0%BB%D0%B0-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%BA%D0%B8-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B9-%D0%B2-%D0%BE%D1%82%D0%B4%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%B9-%D0%BA%D0%BB%D0%B0%D1%81%D1%81

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,16 +1,24 @@
 
-    public static void main(String... args) throws IOException {
         // считываем состояние дома из файла
         Gson gson = new Gson();
```
Чтение - в отдельный класс, нарушение SRP.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -76,5 +42,16 @@ private static SensorEvent getNextSensorEvent() {
+        ApplicationContext context = new AnnotationConfigApplicationContext(MainConfiguration.class);
+        SensorEventsManager sensorEventsManager = context.getBean(SensorEventsManager.class);
+        sensorEventsManager.start();
+>>>>>>> Stashed changes
```
`======= ... >>>>>>> Stashed changes` - код не компилируется.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Bean
+    EventProcessor alarmEventProcessor() {
+        return new AlarmProcessor();
```
Если вы берете, как сказали, чужой код, так берите все классы.
Где код класса AlarmProcessor?

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Light.java:
```diff
@@ -20,4 +22,16 @@ public String getId() {
     }
+
+    public void handleEvent(SensorEvent event, String roomName) {
+        if (this.getId().equals(event.getObjectId())) {
```
Делегируйте обработку событий отдельным классам https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%B4%D0%B5%D0%BB%D0%B5%D0%B3%D0%B8%D1%80%D1%83%D0%B9%D1%82%D0%B5-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%BA%D1%83-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B9-%D0%BE%D1%82%D0%B4%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%BC-%D0%BA%D0%BB%D0%B0%D1%81%D1%81%D0%B0%D0%BC

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Bean
+    EventProcessor hallDoorEventProcessor() {
+        return new HallDoorProcessor();
```
Где код класса HallDoorProcessor?

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Bean
+    EventProcessor doorEventProcessor() {
+        return new DoorProcessor();
```
Где код класса DoorProcessor?

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Bean
+    EventProcessor lightEventProcessor() {
+        return new LightProcessor();
```
Где код класса LightProcessor?

a-ivanov reviewed (changes requested):
Бинарные классы необходимо удалить из репозитория. Это касается всех *.jar. В репозитории - только файлы с кодом *.java (в ресурсах может лежать *.js).
Нет кода по 3 дз - сигнализации и декоратора, отправщика смс.

a-ivanov commented:
Перечитайте задание 2, не вижу код обработки Action в объектах дома.

a-ivanov commented:
> Был вынужден во многом обращаться за "вдохновением" к чужим кодам. Подумал, что лучше так и выполнить задание, чем не выполнить вообще. Если вас обеспокоит "похожесть" кода, смело вините только меня.
> Большую часть этих изменений я закоммитил ещё неделю назад, но закоммитил в ветку hometask-1, поэтому вы не увидели этих изменений(можно в истории посмотреть). В этой ветке первым коммитом откатил изменения назад и следующим добавил изменения(домашки 2, 3 и 4)

Судя по вашему коду складывается впечатление, что вы сами не поняли, что вы скопировали.
Чего конкретно непонятно по 2-4 заданиям?

anton-kuvaldin committed:
Доработки

Исправлены замечания

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Bean
+    public  HashMap<String, SensorEventType> getAdapter(){
```
Бины объявляйте типом интерфейса `Map<String, SensorEventType>`.

a-ivanov replied:
Клиент не должен знать детали реализации, чтобы было легко ее менять в конфигурации.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Bean
+    public EventProcessorAdapter doorsEventProcessor(SmartHome smartHome, HashMap<String, SensorEventType> adapter){
```
Тут тоже `Map<String, SensorEventType> adapter`.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/DoorEventProcessor.java:
```diff
@@ -0,0 +1,29 @@
+import static ru.sbt.mipt.oop.SensorEventType.DOOR_OPEN;
+
+public class DoorEventProcessor implements EventProcessor {
+    private final CommandSender commandSender = new CommandSenderImpl();
```
В аргумент конструктора DoorEventProcessor.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
         return rooms;
     }
+
+    public void handleEvent(Action action) {
```
Название метода handle и event, а аргумент - Action. Как это можно понять?
Тогда переименуйте processAction() или что-то такое.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
         return rooms;
     }
+
+    public void handleEvent(Action action) {
```
Добавьте интерфейс `Actionable` с методом выполнения `Action`, см. слайды к 2 дз.

a-ivanov replied:
implements Actionable

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
     }
 
+    public void formSignalizationObj() {
+        signalization = new Signaling(PIN);
```
Зачем создавать сигнализацию? Передавайте ее через аргумент в конструкторе. Уже с установленным PIN.
Зачем дому знать детали сигнализации? А если захотите поменять ее реализацию?

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
 import java.util.Collection;
+import java.util.Objects;
 
 public class SmartHome {
```
`implements Actionable`

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/HallDoorEventProcessor.java:
```diff
@@ -0,0 +1,50 @@
+    private final CommandSender commandSender;
+
+    public HallDoorEventProcessor() {
+        this.commandSender = new CommandSenderImpl();
```
Аргумент в конструкторе, передавайте снаружи созданный объект.

a-ivanov replied:
Не сделано.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Signaling.java:
```diff
@@ -0,0 +1,39 @@
+        return pin;
+    }
+
+    public boolean checkPassword(int password){
```
Не надо давать клиентам доступ к sensitive информации, сделайте package private.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SignalingState.java:
```diff
@@ -0,0 +1,10 @@
+
+    void setDeactivated(int PIN);
+
+    void AlarmOn();
```
`alarmOn()`

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
         rooms.add(room);
     }
 
     public Collection<Room> getRooms() {
```
По 2 дз удалите этот метод, все взаимодействие с домом - через Actionable.

a-ivanov replied:
Не сделано

a-ivanov replied:
Ну не сделано же

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,23 +2,49 @@
+    public void formSignalizationObj() {
+        signalization = new Signaling(PIN);
+    }
     public void addRoom(Room room) {
```
По 2 дз удалите этот метод, все взаимодействие с домом - через Actionable.

a-ivanov replied:
Не сделано

a-ivanov replied:
Ну не сделано же

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SignalingDeactivatedState.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public void setActivated(int PIN) {
+        if (PIN == signalization.pin) {
```
Здесь просто активируйте без проверки пароля (по типу, как работает установка кода в сейфе в отеле).

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Signaling.java:
```diff
@@ -0,0 +1,39 @@
+        return signalizationState;
+    }
+
+    public void changeState(SignalingState signalizationState){
```
package private, чтобы клиенты не могли ничего сделать/не знали о состояниях, это деталь реализации.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Signaling.java:
```diff
@@ -0,0 +1,39 @@
+        this.pin = PIN;
+    }
+
+    public SignalingState getSignalizationState() {
```
Удалите этот метод, не показывайте деталь реализации. При необходимости сменить реализацию с шаблона состояние, придется менять много клиентского кода.

a-ivanov replied:
Не сделано.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Signaling.java:
```diff
@@ -0,0 +1,39 @@
+        this.signalizationState = signalizationState;
+    }
+
+    private int getCode() {
```
Не надо давать клиентам доступ к sensitive информации, сделайте package private.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,73 @@
+        return new EventProcessorAdapter(new HallDoorEventProcessor(), smartHome, adapter);
+    }
+    @Bean
+    public EventProcessorAdapter lightEventProcessor(SmartHome smartHome, HashMap<String, SensorEventType> adapter){
```
Объявите адаптер типом EventHandler из сторонней библиотеки, чтобы SensorEventsManager не знал, что вы используете адаптер. Так будет легче менять / добавлять реализацию обработки событий.

a-ivanov reviewed (changes requested):
Где реализация декоратора для сигнализации, отправка смс из 3 дз?

anton-kuvaldin committed:
Доработки 2

Не осилил все правки, так как упёрся в StackOverFlow и не смог с ним разобраться. Частично сделал RemoteControl. Всё работает, если оставить метод formSignalizationObj() в SmartHome

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -2,16 +2,25 @@
 
     public SmartHome(Collection<Room> rooms) {
         this.rooms = rooms;
+        this.signalization = new Signaling();
```
Это вынести в аргумент конструктора.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SignalingEventProcessor.java:
```diff
@@ -0,0 +1,13 @@
+    @Override
+    public void processEvent(SmartHome smartHome, SensorEvent sensorEvent) {
+        if (sensorEvent.getType() == SensorEventType.SIGNALIZATION_ACTIVATED) {
+            smartHome.getSignaling().getSignalizationState().setActivated(sensorEvent.getPIN());
```
Состояния должны менять сами состояния через публичные методы в сигнализации. Доступ к методу по смене состояния необходимо сделать package private, это деталь реализации и о ней должны знать только сигнализация и  классы состояний

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/AdapterNewApi.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public void handleEvent(CCSensorEvent event) {
+        SensorEvent sensorEvent = new SensorEvent(mapper.get(event.getEventType()), event.getObjectId());
+        sensorEvent.setPIN(666);
```
Почему у каждого события есть код? Это нужно только для событий сигнклизации

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/AlarmSendMessageDecorator.java:
```diff
@@ -0,0 +1,29 @@
+                }
+                Signaling signaling = (Signaling) object;
+                if (signaling.isAlarmed() || signaling.isActivated()) {
+                    System.out.println("Send SMS!");
```
Это нарушение SRP, создайте интерфейс Notifier с имплементацией DummySMSNotifier. Внутри реализации печатайте отправку смс. Здесь же создайте поле типа Notifier, передайте через конструктор нужную реализацию

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Bean
+    Map<String, SensorEventType> getAdapter() {
```
Не надо использовать глаголы в названии бинов, кроме этого называйте по типу создаваемого объекта, например eventTypeMap

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MainConfiguration.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Bean
+    public AdapterNewApi signalizationEventAdapter(SmartHome smartHome) {
```
Лучше объявлять бины типами имплементируемого ими интерфейса, чтобы было легче менять имплементацию.

anton-kuvaldin committed:
Доработки 3

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Signaling.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public boolean checkPassword(int PIN){
+        return PIN == this.getCode();
```
Можно же просто обращаться к полю напрямую.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/HallDoorEventProcessor.java:
```diff
@@ -0,0 +1,51 @@
+    private final CommandSender commandSender;
+
+    public HallDoorEventProcessor(CommandSender commandSender) {
+        //this.commandSender = new CommandSenderImpl();
```
Удаляйте закомментированный код

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/RCConfig.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Bean
+    CommandRC ActivateSignalizationCommand(SmartHome smartHome) {
```
С маленькой буквы название бина
